### PR TITLE
feat(cake_kda): add serving-native packed decode backend

### PR DESCRIPTION
**Pull request by yyihuang**
## Description

This PR adds a serving-native CAKE KDA decode backend behind a new independent
`flashinfer.packed_kda_decode` API. It consumes post-convolution packed QKV,
raw gate and beta logits, and an indexed BF16 recurrent-state pool directly,
so serving integrations do not need to split/copy Q/K/V, preprocess gate/beta,
or gather/scatter state around the recurrent update.

This API is intentionally separate from `recurrent_kda` (already-split inputs)
and `fused_kda_decode` (which also owns convolution and gated RMSNorm).

### Exact contract

- `T=1`, `H=HV=12`, `K=V=128`, BF16 data/state;
- packed QKV `[B, 4608]`, raw gate `[B, 1536]`, raw beta `[B, 12]`;
- FP32 `A_log[12]` and `dt_bias[1536]`;
- indexed state `[N, 12, 128, 128]`, with compact inner dimensions and an
  arbitrary positive/disjoint outer slot stride;
- contiguous int32 state indices; unique in-bounds active slots are a caller
  contract, and `-1` is an inactive graph-padding row;
- fixed `scale=1/sqrt(128)`, L2 epsilon `1e-6`, and `lower_bound=-5` semantics;
- canonical caller-owned output `[B, 1, 12, 128]` and the caller's current
  PyTorch CUDA stream.

The public call is strict: unsupported devices or tensor contracts raise and
never silently route to another implementation.

### Frozen schedules and architecture boundary

- `B < 32`: tile8 schedule, grid `(192, B, 1)`;
- `B >= 32`: tile16 schedule, grid `(96, B, 1)`;
- block `(32, 1, 1)`, zero dynamic shared memory;
- exact `sm_100a` and exact `sm_103a` JIT/AOT modules; no `sm_100f` claim.

The frozen generated bodies are protected by raw-body SHA-256 tests:

- tile8: `d0de8869242d09bf0c1c4840a7fd73dcd32835050cdc08db58b19a2c7506d0da`;
- tile16: `d8a446e42da47e2d8cd05139c77efe9c970f2d36394b68b49649beb6bc2bbfbe`.

### Integration and tests

- independent TVM-FFI binding and target-aware JIT/AOT registration;
- public API, top-level export, trace definition including the in-place state
  side effect, and API documentation;
- selector, frozen-source, binding, and AOT tests;
- GPU correctness coverage for B=1/8/16/31/32/64/128/256/512, production
  strides, non-identity and `-1` indices, untouched state/padding, current
  stream, CUDA Graph changed-input/index replay, and a 512-step FP64 diagnostic;
- strict CUPTI direct/graph benchmark with clean-source and loaded-binary
  digests.

## Validation

PR head: `826ed5f2fdc301440aa1ec799066f2fa3dfb63d5`.

The performance-evidence commit is its direct parent
`89a5d3fcee8f8ddcff9da5547d095ffd73c1ef79`. The follow-up only pre-includes
`math_constants.h` outside the generated-type rename boundary and sorts a
Python `__all__`; it does not change either frozen body, the binding checks,
launch parameters, or public API. The exact PR head was revalidated on both
architectures after that review follow-up.

- focused CPU/JIT/AOT/facade tests: 27 passed;
- pre-commit on all changed files: passed;
- frozen tile8/tile16 body hashes: exact match.

### Exported public-API correctness

| Target | GPU | JIT tests | GPU tests | Worst output max abs | Worst state max abs | Sanitizer |
|---|---|---:|---:|---:|---:|---|
| `sm_100a` | NVIDIA B200 | 19 passed | 26 passed | `1.5259e-5` | `2.4414e-4` | synccheck + memcheck: 0 errors |
| `sm_103a` | NVIDIA GB300 | 19 passed | 26 passed | `7.6294e-6` | `4.8828e-4` | synccheck + memcheck: 0 errors |

Both targets cover the complete batch sweep, production strides, non-identity
and inactive indices, current-stream launches, CUDA Graph changed-input replay,
and the 512-step FP64 diagnostic.

### Exported public API vs exact SGLang packed Triton

Median GPU span from strict CUPTI, cold L2, 30 rounds per backend, same inputs,
and alternating AB/BA order. The baseline is the exact raw SGLang packed Triton
serving kernel (source SHA-256 `cf5e980d86174a631bcd0872f8ebf7a6ab7c21c3435c097f66f8ba0e686debc0`),
not a copied CAKE-internal timing table.

#### B200 / `sm_100a`

| B | Direct CAKE / Triton (us) | Direct speedup | CUDA Graph CAKE / Triton (us) | Graph speedup |
|---:|---:|---:|---:|---:|
| 1 | 3.6800 / 4.7360 | 1.2870x | 3.8395 / 4.8160 | 1.2543x |
| 8 | 4.5120 / 5.5355 | 1.2268x | 4.7035 / 5.4720 | 1.1634x |
| 16 | 5.6960 / 6.5275 | 1.1460x | 5.8875 / 6.5440 | 1.1115x |
| 31 | 8.4480 / 8.7360 | 1.0341x | 8.6400 / 8.6720 | 1.0037x |
| 32 | 8.6885 / 8.7840 | 1.0110x | 8.8635 / 8.7675 | 0.9892x |
| 64 | 13.7280 / 15.1525 | 1.1038x | 13.8080 / 15.2000 | 1.1008x |
| 128 | 22.2880 / 24.2240 | 1.0869x | 22.2080 / 24.5120 | 1.1037x |
| 256 | 39.8555 / 41.9840 | 1.0534x | 40.0000 / 41.7440 | 1.0436x |
| 512 | 75.0550 / 77.5995 | 1.0339x | 75.0240 / 77.3760 | 1.0313x |
| **Geomean** | | **1.1058x** | | **1.0863x** |

#### GB300 / `sm_103a`

| B | Direct CAKE / Triton (us) | Direct speedup | CUDA Graph CAKE / Triton (us) | Graph speedup |
|---:|---:|---:|---:|---:|
| 1 | 4.0160 / 4.6080 | 1.1474x | 4.0645 / 4.4640 | 1.0983x |
| 8 | 4.5760 / 5.3600 | 1.1713x | 4.6720 / 5.3760 | 1.1507x |
| 16 | 5.6485 / 6.4000 | 1.1330x | 5.8080 / 6.4000 | 1.1019x |
| 31 | 8.7040 / 8.1600 | 0.9375x | 8.8800 / 8.1920 | 0.9225x |
| 32 | 8.3525 / 8.2560 | 0.9884x | 8.5280 / 8.2560 | 0.9681x |
| 64 | 13.2480 / 14.3520 | 1.0833x | 13.4400 / 14.1920 | 1.0560x |
| 128 | 21.0240 / 22.8320 | 1.0860x | 21.1200 / 23.0565 | 1.0917x |
| 256 | 37.7125 / 40.2400 | 1.0670x | 37.9675 / 40.1760 | 1.0582x |
| 512 | 70.4320 / 73.5685 | 1.0445x | 70.6720 / 73.8725 | 1.0453x |
| **Geomean** | | **1.0708x** | | **1.0526x** |

Every timed sample on both GPUs contained exactly one launch and one kernel
activity with zero inter-kernel gap. Across all batches and both modes, worst
output max abs was `1.5259e-5`, worst state max abs was `4.8828e-4`, and
unselected state slots and padding were bitwise preserved. Regressions at
GB300 B31/B32 and B200 graph B32 are retained rather than hidden by geomeans.

A same-configuration cross-run Kimi-K3 SGLang comparison reports CAKE output
throughput `1985.439 tok/s` versus the recorded Triton baseline
`1979.075 tok/s` (`1.0032x`), while median TPOT is effectively flat
(`30.814 ms` versus `30.806 ms`, `0.9998x`). GSM8K medians are `0.975` for
CAKE and `0.980` for Triton, passing the configured quality gate. Because the
measurements use different source snapshots and allocations and the delta is
noise-scale, they do not establish a material E2E speedup. Full runs and
methodology are in the SGLang PR.

SGLang integration: https://github.com/sgl-project/sglang/pull/33647

Related to https://github.com/flashinfer-ai/flashinfer/issues/4254.

### CODE REVIEW

File: benchmarks/bench_packed_kda_decode.py [hunk 1]
```diff
@@ -0,0 +1,517 @@
...
+GATE_WIDTH = HEADS * HEAD_DIM
+PRODUCTION_MIXED_STRIDE = 6144
+PRODUCTION_GATE_STRIDE = GATE_WIDTH
+PRODUCTION_BETA_STRIDE = HEADS
+STATE_SLOT_ELEMENTS = HEADS * HEAD_DIM * HEAD_DIM
+PRODUCTION_STATE_PADDING = 256
+DEFAULT_BATCHES = (1, 8, 16, 31, 32, 64, 128, 256, 512)
+SUPPORTED_ARCHS = {(10, 0): "sm100a", (10, 3): "sm103a"}
+DATA_SEED = 20260805
+TYPED_SOURCE_SHA256 = "24edeaf9676b12ec3301ff413080194282be0b35e604f785dffba41c0c48640e"
```
[the remark this answers is not in the document]
Comment (yyihuang) [reply benchmarks/bench_packed_kda_decode.py:56]: This constant is intentionally the provenance hash of the canonical CAKE Weave typed source that generated the two frozen CUDA bodies; it is not the hash of `flashkda_packed_t1_binding.cuh`. That internal typed source is not shipped in this standalone export, so comparing the value to the local binding would always be incorrect. The locally shipped artifacts are verified fail-closed by `_verify_frozen_body_hashes()`, and the JIT tests also assert that each frozen file carries the same typed-source provenance. I am therefore leaving this as provenance metadata rather than adding the proposed mismatched check.

File: csrc/kda/flashkda_packed_t1_binding.cuh [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+#define uint8_t flashkda_packed_generated_uint8_t
+#define uint16_t flashkda_packed_generated_uint16_t
+#define uint32_t flashkda_packed_generated_uint32_t
+#define uint64_t flashkda_packed_generated_uint64_t
+#define int32_t flashkda_packed_generated_int32_t
+#define int16_t flashkda_packed_generated_int16_t
+#define LoomTensorMap flashkda_packed_generated_LoomTensorMap
+#define LoomTensorMapPack flashkda_packed_generated_LoomTensorMapPack
+#define CUtensorMap flashkda_packed_generated_CUtensorMap
+#include FLASHKDA_PACKED_T1_BODY_FILE
```
[the remark this answers is not in the document]
Comment (yyihuang) [reply csrc/kda/flashkda_packed_t1_binding.cuh:56]: Fixed in `826ed5f2`: `<math_constants.h>` is now included before the fixed-width rename boundary. The exact follow-up commit passed 19 JIT + 26 GPU tests and synccheck/memcheck with 0 errors on both B200 (job 1630062) and GB300 (job 378178).

File: flashinfer/jit/flash_kda_packed_t1.py [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+    "FLASH_KDA_PACKED_T1_VARIANT_METADATA",
+    "FLASH_KDA_PACKED_T1_VARIANTS",
+    "FlashKDAPackedT1Target",
+    "FlashKDAPackedT1Variant",
+    "FlashKDAPackedT1VariantMetadata",
+    "gen_flash_kda_packed_t1_module",
+    "get_flash_kda_packed_t1_module",
+    "get_flash_kda_packed_t1_uri",
+    "load_flash_kda_packed_t1_module",
+]
```
[the remark this answers is not in the document]
Comment (yyihuang) [reply flashinfer/jit/flash_kda_packed_t1.py:246]: Fixed in `826ed5f2`. Pre-commit (including Ruff) passes, and the exact commit also passed the B200/GB300 JIT and GPU validation noted above.

**Comment by ameynaik-hub** [comment]
https://github.com/flashinfer-ai/flashinfer/pull/4417 I have this PR which kda decode kernel based on my kernel gdn decode kernel which is ~10.7% better than this version. 

| B | Direct CAKE (µs) | Direct CuTe (µs) | Direct speedup | Graph CAKE (µs) | Graph CuTe (µs) | Graph speedup |
|---|---|---|---|---|---|---|
| 1 | 4.1600 | 3.9355 | 1.0570× | 4.0965 | 3.9040 | 1.0493× |
| 8 | 4.8640 | 4.6720 | 1.0411× | 4.9600 | 4.7685 | 1.0402× |
| 16 | 6.1280 | 5.8240 | 1.0522× | 6.1440 | 5.8725 | 1.0462× |
| 31 | 8.9600 | 7.6480 | 1.1715× | 9.0240 | 7.7760 | 1.1605× |
| 32 | 9.0240 | 7.7440 | 1.1653× | 9.2160 | 7.9360 | 1.1613× |
| 64 | 14.1760 | 12.3845 | 1.1447× | 14.3040 | 12.5760 | 1.1374× |
| 128 | 22.8160 | 20.3200 | 1.1228× | 22.9120 | 20.4480 | 1.1205× |
| 256 | 40.7675 | 36.5440 | 1.1156× | 40.9280 | 36.7040 | 1.1151× |
| 512 | 76.4165 | 68.9445 | 1.1084× | 76.6725 | 69.1525 | 1.1087× |
| Geomean | | | 1.1078× | | | 1.1034× |

**Methodology:**
- Strict CUPTI GPU-span timing
- Cold L2
- 100 measured rounds per backend and configuration
- 100 configured warmup iterations per round
- Identical inputs

**Comment by yyihuang** [comment]
Superseded by #4445. That PR contains this draft's serving-native packed T=1 decode API and frozen SM100a/SM103a kernels, plus the strided-prefill, indexed-state, native-checkpoint, and later build-boundary hardening work. Closing this draft to keep review on the consolidated PR; this branch and its review history remain available.
